Widgets/Widgets/FeaturedArticleWidget.swift

```diff
@@ -0,0 +1,309 @@
+
+	public var body: some WidgetConfiguration {
+		StaticConfiguration(kind: kind, provider: FeaturedArticleProvider(), content: { entry in
+			FeaturedArticleView(entry: entry)
+		})
+		.configurationDisplayName(FeaturedArticleWidget.LocalizedStrings.widgetTitle)
+		.description(FeaturedArticleWidget.LocalizedStrings.widgetDescription)
+		.supportedFamilies([.systemSmall, .systemMedium, .systemLarge])
+	}
+}
+
+// MARK: - Timeline Entry
+
+struct FeaturedArticleEntry: TimelineEntry {
+
+	// MARK: - Properties
+
+	var date: Date
+	var content: WidgetFeaturedContent?
+	var fetchError: WidgetContentFetcher.FetcherError?
+
+	// MARK: - Computed Properties
+
+	var hasDisplayableContent: Bool {
+		return fetchError == nil && content?.featuredArticle != nil
+	}
+
+	var fetchedLanguageCode: String? {
+		return content?.featuredArticle?.languageCode
+	}
+
+	var title: String {
+		return content?.featuredArticle?.normalizedTitle ?? ""
+	}
+
+	var description: String {
+		return content?.featuredArticle?.description ?? ""
+	}
+
+	var extract: String {
+		return content?.featuredArticle?.extract ?? ""
+	}
+
+	var layoutDirection: LayoutDirection {
+		if let direction = content?.featuredArticle?.languageDirection {
+			return direction == "rtl" ? .rightToLeft : .leftToRight
+		}
+
+		return .leftToRight
+	}
+
+	var contentURL: URL? {
+		guard let page = content?.featuredArticle?.contentURL.desktop.page else {
+			return nil
+		}
+
+		return URL(string: page)
+	}
+
+	var imageData: Data? {
+		return content?.featuredArticle?.thumbnailImageSource?.data
+	}
+
+}
+
+// MARK: - Timeline Provider
+
+struct FeaturedArticleProvider: TimelineProvider {
+	typealias Entry = FeaturedArticleEntry
+
+	func placeholder(in context: Context) -> FeaturedArticleEntry {
+		return FeaturedArticleEntry(date: Date(), content: nil)
+	}
+
+	func getSnapshot(in context: Context, completion: @escaping (FeaturedArticleEntry) -> Void) {
+		WidgetController.shared.fetchFeaturedContent(isSnapshot: context.isPreview) { result in
+			let currentDate = Date()
+			switch result {
+			case .success(let featuredContent):
+				completion(FeaturedArticleEntry(date: currentDate, content: featuredContent))
+			case .failure(let fetchError):
+				completion(FeaturedArticleEntry(date: currentDate, content: nil, fetchError: fetchError))
+			}
+		}
+	}
+
+	func getTimeline(in context: Context, completion: @escaping (Timeline<FeaturedArticleEntry>) -> Void) {
+		WidgetController.shared.fetchFeaturedContent { result in
+			let currentDate = Date()
+			switch result {
+			case .success(let featuredContent):
+				completion(Timeline(entries: [FeaturedArticleEntry(date: currentDate, content: featuredContent)], policy: .after(currentDate.randomDateShortlyAfterMidnight() ?? currentDate)))
+			case .failure(let fetchError):
+				completion(Timeline(entries: [FeaturedArticleEntry(date: currentDate, content: nil, fetchError: fetchError)], policy: .atEnd))
```

Comment:
My intent was to avoid any cases of bad or stale featured article widget data by any means within our control. It’s unknown to me the internal specifics of how Apple uses these policies. I’d like to assume that they use some intelligence, like if the device suddenly changes from the network being unavailable to being available, to help in determining when to execute `.atEnd` policy timelines, but I just don’t know for sure.

Frankly, I’d prefer if we’d transition the other widgets to use a similar strategy as here (unless it has a strong negative effect on Wikimedia network traffic) rather than adopt their strategies here. But again, if you feel strongly about it, lemme know I can add in the delay.